### ISSUE TITLE
Implementation of Beamforming in the driver is dummy

### ISSUE THREAD

**Issue by Vito-Swift** [morrownr/8814au#44]
Hi everyone, I'm working on a project that needs to **read the channel estimation matrix out from the NDP procedure**. 
To do so, I used an EDUP-1622 wireless card (with up to 3 spatial stream supported) and planed to do some minor modifications on rtl8814au driver for my project. 

In the first step, I setuped 2 devices in AP/STA scheme, and confirmed the connection is using 802.11AC standard. 

```
$ sudo lshw -C network
... ...
  *-network:2
       description: Wireless interface
       physical id: 3
       bus info: usb@2:1
       logical name: wlan1
       serial: e8:4e:06:95:28:cd
       capabilities: ethernet physical wireless
       configuration: broadcast=yes driver=rtl8814au driverversion=5.10.63-v7l+ ip=169.254.138.77 multicast=yes wireless=IEEE 802.11AC
```

To better describe the protocol-related parameters of my setup, I pasted my hostapd.conf as below:
```
interface=wlan1
driver=nl80211
ssid=Beamformer
macaddr_acl=0
auth_algs=1
ignore_broadcast_sisd=0
beacon_int=100
dtim_period=1
wpa=2
wpa_passphrase=xxxxxxxx
wpa_key_mgmt=WPA-PSK WPA-PSK-SHA256
wpa_pairwise=CCMP
rsn_pairwise=CCMP

hw_mode=a
ieee80211n=1
ieee80211ac=1
channel=36
wmm_enabled=1
wme_enabled=1
country_code=US

ht_capab=[LDPC][HT40-][HT40+][SHORT-GI-20][SHORT-GI-40][MAX-AMSDU-7935]
vht_capab=[RXLDPC][MAX-MPDU-11454][SHORT-GI-80][SU-BEAMFORMER][SU-BEAMFORMEE][BF-ANTENNA-2][SOUNDING-DIMENSION-2][HTC-VHT]

vht_oper_chwidth=1
vht_oper_centr_freq_seg0_idx=42
```

Next, I checked `dmesg -wH` report which shows that each time the TX initiates an NDP procedure the driver will put a line in the kernel log. 
```
$ sudo dmesg -wH
... ...
[  +4.080076] RTW: [C2H], C2H_TXBF!!
[  +4.080137] RTW: [C2H], C2H_TXBF!!
[  +4.080012] RTW: [C2H], C2H_TXBF!!
[  +4.080075] RTW: [C2H], C2H_TXBF!!
[  +4.080042] RTW: [C2H], C2H_TXBF!!
[  +4.079979] RTW: [C2H], C2H_TXBF!!
[  +4.080090] RTW: [C2H], C2H_TXBF!!
[  +4.080073] RTW: [C2H], C2H_TXBF!!
[Oct23 21:14] RTW: [C2H], C2H_TXBF!!
[  +4.080024] RTW: [C2H], C2H_TXBF!!
[  +4.079886] RTW: [C2H], C2H_TXBF!!
[  +4.080050] RTW: [C2H], C2H_TXBF!!
[  +4.080142] RTW: [C2H], C2H_TXBF!!
[  +4.079968] RTW: [C2H], C2H_TXBF!!
[  +4.080043] RTW: [C2H], C2H_TXBF!!
[  +4.080037] RTW: [C2H], C2H_TXBF!!
[  +4.080112] RTW: [C2H], C2H_TXBF!!
[  +4.080107] RTW: [C2H], C2H_TXBF!
```

However, when I look into the source code to investigate where the channel estimation matrix is stored and how it is being modified by the NDP procedure, I found the implementation seems dummy. The related source code is pasted below 

[source code here:](https://github.com/morrownr/8814au/blob/ee25b01fe530c70f7343c6d7c17d741ff9e48890/hal/phydm/txbf/haltxbfinterface.c#L977)

```c
#elif (DM_ODM_SUPPORT_TYPE == ODM_CE)

u32 beamforming_get_report_frame(
	void *dm_void,
	union recv_frame *precv_frame)
{
	struct dm_struct *dm = (struct dm_struct *)dm_void;
	u32 ret = _SUCCESS;
	struct _RT_BEAMFORMEE_ENTRY *beamform_entry = NULL;
	u8 *pframe = precv_frame->u.hdr.rx_data;
	u32 frame_len = precv_frame->u.hdr.len;
	u8 *TA;
	u8 idx, offset;

	/*@Memory comparison to see if CSI report is the same with previous one*/
	TA = get_addr2_ptr(pframe);
	beamform_entry = phydm_beamforming_get_bfee_entry_by_addr(dm, TA, &idx);
	if (beamform_entry->beamform_entry_cap & BEAMFORMER_CAP_VHT_SU)
		offset = 31; /*@24+(1+1+3)+2  MAC header+(Category+ActionCode+MIMOControlField)+SNR(nc=2)*/
	else if (beamform_entry->beamform_entry_cap & BEAMFORMER_CAP_HT_EXPLICIT)
		offset = 34; /*@24+(1+1+6)+2  MAC header+(Category+ActionCode+MIMOControlField)+SNR(nc=2)*/
	else
		return ret;

	return ret;
}
```

To my understanding of the 802.11ac vht standard and this driver, when each time beamformer initials a NDP procedure and the beamformee responds with a compressed channel matrix, the beamformer shall enter this `beamforming_get-report_frame()` function to collect the compressed channel matrix and store it in the `beamforming_entry->csi_matrix`. But in the implementation, ***I did not see the csi_matrix being modified***. 

Also, I print the values of the csi_matrix in the kernel long after the connection is established, but it seems to me the csi_matrix stays unchanged (the value of all the elements in csi_matrix[] are zeros and will never changed).

Is that mean this driver has not implemented beamforming functionality yet?

**Comment by libc0607** [comment]
Hi @Vito-Swift ,

I'm currently looking into the beamforming implementation in Realtek's driver and Google brought me here. My goal is to use its beamforming feature in FPV (bi-directional injecting & monitoring, without establishing AP/STA connection).

> From the source code, I could find some clue that the beamforming functionality is implemented in the hardware instead of the software, because most beamforming related commands are called through the hardware abstract layer (rtw_bf_hal). After printing out the content of memory, I suspect the beamforming capabilities are correctly implemented in the firmware, so maybe I could close this issue at this time. 

I did notice the ```_send_*ht_ndpa_packet()``` in ```rtw_beamforming.c``` triggered by some ```rtw_bf_cmd```, but I haven't noticed any code related to receiving NDPA packet or sending NDP packet (maybe when acting as a SU beamformee?). Does that mean -- to start a beamforming process, only need to config some regs, send an NDPA packet (in the driver), then the rest is automatically done by hardware or firmware?

**Comment by Vito-Swift** [comment]
> Hi @Vito-Swift ,
> 
> I'm currently looking into the beamforming implementation in Realtek's driver and Google brought me here. My goal is to use its beamforming feature in FPV (bi-directional injecting & monitoring, without establishing AP/STA connection).
> 
> > From the source code, I could find some clue that the beamforming functionality is implemented in the hardware instead of the software, because most beamforming related commands are called through the hardware abstract layer (rtw_bf_hal). After printing out the content of memory, I suspect the beamforming capabilities are correctly implemented in the firmware, so maybe I could close this issue at this time.
> 
> I did notice the `_send_*ht_ndpa_packet()` in `rtw_beamforming.c` triggered by some `rtw_bf_cmd`, but I haven't noticed any code related to receiving NDPA/NDP packet or sending the matrix (maybe when acting as a SU beamformee?). Does that mean -- to start a beamforming process, only need to config some regs, send an NDPA packet (in the driver), then the rest is automatically done by hardware or firmware?


I guess the whole channel sounding process is done in the hardware/firmware so there is no clear implementation in the code of this driver. One reason I guess is that the time interval between NDPA and NDP is really short according to the 802.11 standard, and another is that realtek just decided not to disclose this part. 

I have no idea whether injecting NDPA/NDP can be done with some firmware commands instead of using monitor interfaces. Others are welcome to comment on this part.

**Comment by libc0607** [comment]
@Vito-Swift Thank you for your reply. 

> One reason I guess is that the time interval between NDPA and NDP is really short according to the 802.11 standard and putting an extra control logic at the software level does not make sense. 

It does make sense. I forgot this part. 

I've been reading the [rtw88's beamformee implementation in mainline Linux](https://github.com/torvalds/linux/blob/master/drivers/net/wireless/realtek/rtw88/bf.c) (because the Realtek driver is a mess) and I recognized that the beamforming is most likely implemented in hardware. There are register-level configurations, such as MAC address and NDP RX timer.  

Gonna explore more on Realtek's driver, and will update if something new discovered.

> I have no idea whether injecting NDPA/NDP can be done with some firmware commands instead of using monitor interfaces.  

I think the NDPA packet is assembled and sent at [here](https://github.com/morrownr/8814au/blob/b5a6f96cd20a1c6a35ccb000f2f4924dd19704a3/core/rtw_beamforming.c#L808) -- it's similar to monitor injection. The MU report poll is similar, at [here](https://github.com/morrownr/8814au/blob/b5a6f96cd20a1c6a35ccb000f2f4924dd19704a3/core/rtw_beamforming.c#L486). 

It seems that the hardware do send the NDP packet. Maybe triggered by recognizing a TX packet with an "NDPA" type set on the Frame Control field? 

> Others are welcome to comment on this part.

Me too.

**Comment by libc0607** [comment]
Hi @Vito-Swift, updates -- I've done some experiments on rtl8812eu. 

> I guess the whole channel sounding process is done in the hardware/firmware so there is no clear implementation in the code of this driver. 

That's true. 

> I have no idea whether injecting NDPA/NDP can be done with some firmware commands instead of using monitor interfaces. Others are welcome to comment on this part.

Now I can inject VHT NDPA & NDP, without establishing any AP/STA connection, to any Receiver Address with any Transmitter Address, by calling _send_*ht_ndpa_packet(). I'll push the code to my 8812eu repo later. 

I didn't find any firmware command, instead, the whole process is done as follows: (1) setting some register, (2) manually sending the VHT NDPA (set in TX descriptor) frame, then (3) the NDP frame will automatically be sent (by firmware/hardware) after the NDPA. And (4) bfee receives the NDPA & NDP, the Compressed Beamforming Report frame was automatically calculated & sent by beamformee adapter's firmware/hardware. 

The image below is the voltage of the adapter's power amplifier's VDET output (indicates transmit status). The first peak is injected NDPA, followed by a SIFS slot, then the second peak stands for NDP. The same pattern followed was automatic retransmission (as no bfee is active at the moment). Seems all timing meets the standard.
![image](https://github.com/user-attachments/assets/23b867cb-1517-4a8b-81ca-a453f6b7f087)

The image below is one CBR frame captured on the beamformer side, with 2 8812eu (2T2R) as bfer & bfee. So it seems that the beamforming process has been correctly implemented in 8812eu hardware/firmware too.  
<img width="750" alt="bdc4c8e9c28d91def7a8fbea170e252" src="https://github.com/user-attachments/assets/16e3eb04-ee57-4186-b483-a33fc2e521fc">

The code in the driver is mainly a finite-state machine for keeping the information of associated AP/STAs up-to-date. (Why Realtek had implemented >= 3 versions of the beamforming state machine in the newer driver?)

BTW I'm confused: how can I check if the beamforming (the V matrix) really applied to transmission? For a 2TX adapter, there's no obvious gain can be seen ...

**Comment by morrownr** [comment]
@libc0607 @Vito-Swift 

FYI: The project to add rtl8812au and rtl8821au support to the kernel is in the review stage with v2 submitted Friday. You can see it at:

https://patchwork.kernel.org/project/linux-wireless/list/

Look for the patch series:

wifi: rtw88: Add support for RTL8821AU and RTL8812AU

Hopefully these drivers will be accepted soon. My point here is that the original plan was to also add an in-kernel driver for the rtl8814au at some point following the current additions. The work on rtw88 has taken a lot of time, most of which has been by @dubhater. He would have to address whether he still has the time or desire to continue with rtl8814au. I hope he does but that is up to him. Is it possible that the two of you would be interested in assisting if that works for @dubhater ?

**Comment by dubhater** [comment]
It is still the plan to add support for RTL8814AU. I already populated rtw8814au_table.c, which is the most boring part.

**Comment by morrownr** [comment]
@dubhater 

That sounds great. When you are ready for me to help, let me know.

While I have you, I assume you expect me to roll in with some patches once Ping-Ke has accepted the rtl8821au and rtl8812au patch series. Let me know when I need to engage.

**Comment by libc0607** [comment]
Thank both of you for your work in maintaining & porting the drivers!

After looking into Realtek's driver, I've realized how hard and messy it can be... so I truly appreciate your effort

**Comment by morrownr** [comment]
@libc0607 

> After looking into Realtek's driver, I've realized how hard and messy it can be...

Oh, yes and this specific driver has been a real challenge to keep going. Many of the internals in this driver appear to be very old code that is not aging well. I have had to rewrite some of the code just to be able to apply some of the API updates. I decided to basically just freeze the code except for API updates about 3 years ago because this driver is hard to work on given my level of wifi code knowledge. dubhater is much more knowledgeable. There are a lot of Linux users that have adapters based on the rtl8814au chip so hopefully sometime next year we will be able to get the driver into the kernel.

I just hope that the firmware is in decent shape. This project is going to be a challenge and we will needed testers.

Cheers
